http-server-netty/src/main/java/io/micronaut/http/server/netty/body/ImmediateMultiObjectBody.java

```diff
@@ -0,0 +1,140 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.server.netty.body;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.http.server.exceptions.InternalServerException;
+import io.micronaut.http.server.netty.FormRouteCompleter;
+import io.micronaut.http.server.netty.MicronautHttpData;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.ByteBufHolder;
+import io.netty.buffer.ByteBufInputStream;
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.util.ReferenceCountUtil;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Flux;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Immediate {@link MultiObjectBody}, all operations are eager.
+ */
+@Internal
+public final class ImmediateMultiObjectBody extends ManagedBody<List<?>> implements MultiObjectBody {
+    ImmediateMultiObjectBody(List<?> objects) {
+        super(objects);
+    }
+
+    @Override
+    void release(List<?> value) {
+        value.forEach(ReferenceCountUtil::release);
+    }
+
+    public ImmediateSingleObjectBody single(Charset defaultCharset, ByteBufAllocator alloc) {
+        List<?> objects = prepareClaim();
+        if (objects.isEmpty()) {
+            return next(new ImmediateSingleObjectBody(null));
+        }
+        boolean allFormData = true;
+        for (Object object : objects) {
+            if (!(object instanceof MicronautHttpData<?>)) {
+                allFormData = false;
+                break;
+            }
+        }
+        if (allFormData) {
+            //noinspection unchecked
+            return next(new ImmediateSingleObjectBody(toMap(defaultCharset, (List<? extends MicronautHttpData<?>>) objects)));
+        }
+        if (objects.size() == 1) {
+            Object o = objects.get(0);
+            return next(new ImmediateSingleObjectBody(o instanceof ByteBufHolder bbh ? bbh.content() : o));
+        }
+        return next(new ImmediateSingleObjectBody(coerceToComposite(objects, alloc)));
+    }
+
+    private static CompositeByteBuf coerceToComposite(List<?> objects, ByteBufAllocator alloc) {
+        CompositeByteBuf composite = alloc.compositeBuffer();
+        for (Object object : objects) {
+            composite.addComponent(true, ((ByteBufHolder) object).content());
+        }
+        return composite;
+    }
+
+    public static Map<String, Object> toMap(Charset charset, Iterable<? extends MicronautHttpData<?>> dataList) {
+        Map<String, Object> singleMap = new LinkedHashMap<>();
```

Comment:
You might want to use CollectionUtils.newLinkedHashMap to create a map of size dataList.size